### PR TITLE
Refactor callback and message in cloud source file

### PR DESCRIPTION
**Pull request by ramonhpr**

### CODE REVIEW

File: src/cloud.c [hunk 1]
```diff
@@ -58,27 +58,32 @@
...
-	enum {
-		UPDATE_EVT,
-		REQUEST_EVT,
-		ADDED_EVT,
-		REMOVED_EVT,
-	} event_id;
-};
+cloud_cb_t cloud_cb;
+
+static void cloud_msg_destroy(struct cloud_msg *msg)
```
[the remark this answers is not in the document]
Comment (ramonhpr) [reply src/cloud.c]: Done. I removed this function

File: src/cloud.c [hunk 2]
```diff
@@ -58,27 +58,32 @@
...
+{
+	if (msg->type == UNREGISTER_MSG || msg->type == REGISTER_MSG)
+		return;
+
+	if (l_queue_isempty(msg->list))
```
[the remark this answers is not in the document]
Comment (ramonhpr) [reply src/cloud.c]: Indeed. Done

File: src/cloud.c [hunk 3]
```diff
@@ -91,74 +96,55 @@ static bool on_cloud_receive_message(const char *exchange,
...
-	case ADDED_EVT:
-		cb_handler.added_cb = evt_handler->cb;
-		id = parser_get_key_str_from_json_obj(jso, "id");
-		token = parser_get_key_str_from_json_obj(jso, "token");
-		if (!id || !token)
+	case REGISTER_MSG:
+		msg.device_id = parser_get_key_str_from_json_obj(jso, "id");
+		msg.token = parser_get_key_str_from_json_obj(jso, "token");
+		if (!msg.device_id || !msg.token)
 			return false;
```
[the remark this answers is not in the document]
Comment (ramonhpr) [reply src/cloud.c]: Fixed

File: src/cloud.c [hunk 4]
```diff
@@ -257,21 +243,15 @@ int cloud_publish_data(const char *id, uint8_t sensor_id, uint8_t value_type,
 }
 
 /**
- * cloud_set_read_handlers:
- * @on_update: callback to be called when receive an update message
- * @on_request: callback to be called when receive an request message
- * @on_removed: callback to be called when device is removed
+ * cloud_set_read_handler:
+ * @cb: handler callback to handle message received from cloud
```
[the remark this answers is not in the document]
Comment (ramonhpr) [reply src/cloud.c]: Done

File: src/msg.c [hunk 5]
```diff
@@ -1250,8 +1250,7 @@ static void forget_if_unknown(struct knot_device *device, void *user_data)
 		hal_log_info("Removing proxy for %s", id);
 }
 
-static void on_device_added(const char *device_id, const char *token,
-			    void *user_data)
+static bool on_device_added(const char *device_id, const char *token)
```
[the remark this answers is not in the document]
Comment (ramonhpr) [reply src/msg.c]: Done

File: src/msg.c [hunk 6]
```diff
@@ -1337,7 +1333,7 @@ static void unregister_callback(struct l_timeout *timeout, void *user_data)
 	device_forget_destroy(mydevice);
 }
 
-static void on_device_removed(const char *device_id, void *user_data)
+static bool on_device_removed(const char *device_id)
```
[the remark this answers is not in the document]
Comment (ramonhpr) [reply src/msg.c]: Done

File: src/msg.c [hunk 7]
```diff
@@ -1455,42 +1452,43 @@ static void append_on_request_list(void *data, void *user_data)
 				  l_memdup(msg, sizeof(*msg)));
 }
 
-static bool on_update(const char *id, struct l_queue *list,
-		      void *user_data)
+static bool on_cloud_downstream(const char *device_id, struct l_queue *list,
```
[the remark this answers is not in the document]
Comment (ramonhpr) [reply src/msg.c]: Done

File: src/cloud.c [hunk 8]
```diff
@@ -73,76 +81,89 @@ static int map_routing_key_to_msg_type(const char *routing_key)
...
+
+	jso = json_tokener_parse(body);
+	if (!jso) {
+		hal_log_error("Error on parse JSON object");
+		return NULL;
+	}
+
+	msg = create_msg(routing_key, jso);
+	if (!msg)
+		goto free_jso;
```
Comment (alla3) [inline src/cloud.c]: Is this goto necessary? It makes more sense to me to check if the msg is not NULL and execute the following code.
`consumed = cloud_cb(msg, user_data);
cloud_msg_destroy(msg);`

What do you think?

Comment (ramonhpr) [reply src/cloud.c]: Done

**Pull request merged**